<launch>
  <!-- Arguments -->
    <arg name="robot_name" default="ugv"/>
    <arg name="model" default="waffle" doc="model type [burger, waffle, waffle_pi]"/>
    <arg name="x" default="0.0"/>
    <arg name="y" default="0.0"/>
    <arg name="z" default="0.0"/>
    <arg name="node_type" default="ugv"/>
    <arg name="rabbitmq_endpoint" default="localhost"/>
    <arg name="rabbitmq_username" default="admin"/>
    <arg name="rabbitmq_password" default="passw123"/>
    <arg name="rabbitmq_port" default="1883"/>
    <arg name="base_directory" default="$(find multirobot_sim)/scripts"/>
    <arg name="secret" default="abc123"/>
  <!--Robot namespace-->
  <group ns="$(arg robot_name)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />
    <!-- Spawn robot-->
    <node pkg="gazebo_ros" type="spawn_model" name="spawn_urdf"  args="-urdf -model $(arg robot_name) -x $(arg x) -y $(arg y) -z $(arg z) -param robot_description" />
    <!-- Load joint state controller -->
    <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher">
      <param name="use_gui" value="true"/>
      <param name="rate" value="50"/>
      <param name="tf_prefix" value="$(arg robot_name)" />
    </node>
    <!-- Load robot state publisher -->
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg robot_name)" />
    </node>
    <!-- Load the controller -->
    <node pkg="multirobot_sim" type="simple_controller.py" name="simple_controller" output="screen" required="true">
      <param name="odom_topic" value="/$(arg robot_name)/odom"/>
      <param name="cmd_topic" value="/$(arg robot_name)/cmd_vel"/>
      <param name="goal_topic" value="/$(arg robot_name)/goal"/>
      <param name="path_topic" value="/$(arg robot_name)/path"/>
    </node>  
    <node pkg="multirobot_sim" type="roschain.py" name="roschain" output="screen" required="true">
        <param name="node_id" value="$(arg robot_name)"/>
        <param name="node_type" value="$(arg node_type)"/>
        <param name="rabbitmq_endpoint" value="$(arg rabbitmq_endpoint)"/>
        <param name="rabbitmq_username" value="$(arg rabbitmq_username)"/>
        <param name="rabbitmq_password" value="$(arg rabbitmq_password)"/>
        <param name="rabbitmq_port" value="$(arg rabbitmq_port)"/>
        <param name="secret" value="$(arg secret)"/>
        <param name="base_directory" value="$(arg base_directory)"/>
    </node>  
    <node pkg="multirobot_sim" type="task_allocation.py" name="task_allocator" output="screen" required="true">
        <param name="node_id" value="$(arg robot_name)"/>
        <param name="node_type" value="$(arg node_type)"/>
        <param name="odom_topic" value="/$(arg robot_name)/odom"/>
        <param name="update_interval" value="5"/>
    </node>  
  </group>
</launch>
